lib/logstash/filters/multiline.rb

```diff
@@ -246,35 +243,66 @@ def previous_filter!(event, match)
 
   def next_filter!(event, match)
     key = event.sprintf(@stream_identity)
-
-    # protect @pending for race condition between the flush thread and the worker thread
-    pending = @pending[key]
+    pending = @pending[key] ||= []
 
     if match
+      # this line is part of a multiline event, the next line will be part, too, put it into pending.
       event.tag(MULTILINE_TAG)
-      # this line is part of a multiline event, the next
-      # line will be part, too, put it into pending.
-      if pending
-        pending.append(event)
-      else
-        @pending[key] = event
-      end
+      pending << event
       event.cancel
     else
-      # if we have something in pending, join it with this message
-      # and send it. otherwise, this is a new message and not part of
-      # multiline, send it.
-      if pending
-        pending.append(event)
-        event.overwrite(pending)
-        @pending.delete(key)
+      # if we have something in pending, join it with this message and send it.
+      # otherwise, this is a new message and not part of multiline, send it.
+      unless pending.empty?
+        event.overwrite(merge(pending << event))
+        pending.clear
       end
     end # if match
   end
 
-  def collapse_event!(event)
-    event["message"] = event["message"].join("\n") if event["message"].is_a?(Array)
-    event.timestamp = event.timestamp.first if event.timestamp.is_a?(Array)
-    event
+  # merge a list of events. @timestamp for the resulting merged event will be from
+  # the "oldest" (events.first). all @source fields will be deduplicated depending
+  # on @allow_duplicates and joined with \n. all other fields will be deduplicated.
+  # @param events [Array<Event>] the list of events to merge
+  # @return [Event] the resulting merged event
+  def merge(events)
+    dups_key = @allow_duplicates ? @source : nil
+
+    data = events.inject({}) do |result, event|
+      self.class.event_hash_merge!(result, event.to_hash_with_metadata, dups_key)
+    end
+
+    # merged event @timestamp is from first event in sequence
+    data["@timestamp"] = Array(data["@timestamp"]).first
+    # collapse all @source field values
+    data[@source] = Array(data[@source]).join("\n")
+    LogStash::Event.new(data)
   end
+
+  # merge two events data hash, src into dst and handle duplicate values for dups_key
+  # @param dst [Hash] the event to merge into, dst will be mutated
+  # @param src [Hash] the event to merge in dst
+  # @param dups_key [String] the field key to keep duplicate values
+  # @return [Hash] mutated dst
+  def self.event_hash_merge!(dst, src, dups_key = nil)
```

Comment:
this is a strange method for me..the signature of the output changes (for me) unpredictably:

```ruby
> self.event_hash_merge!(dst = { :a => 1 }, src = { :a => 1}); dst
 => {:a=>1} 
> self.event_hash_merge!(dst = { :a => 1 }, src = { :a => 2}); dst
 => {:a=>[1, 2]} 
> # for the example above dst should be { :a => 2 } since I didn't specity :dups_key
```